### PR TITLE
Move default openshift version

### PR DESCRIPTION
**Pull request by mbarnes**
### Which issue this PR addresses:

[[ARO-3896] Read default OCP version from CosmosDB](https://issues.redhat.com/browse/ARO-3896)

Depends on a couple ADO pull requests:
- [RP-Config: Add openShiftVersions data](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config/pullrequest/8603646) :heavy_check_mark: _merged_
- [ARO.Pipelines: Read OpenShiftVersions from YAML configuration](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/ARO.Pipelines/pullrequest/8603681) :heavy_check_mark: _merged_

### What this PR does / why we need it:

Make CosmosDB the single source of truth for available OpenShift versions, including the default version.  Currently it's a mix: CosmosDB has the available versions but only ARO RP knows the default version.

#### Local Development Mode

Local development mode has been a subject of discussion within my functional team (Loki).  Initially I had planned to implement a small hack to keep local development mode working as is in lieu of an internal default OpenShift version and pullspec.  But the team has grander plans for local development mode, so I'm going to defer that issue here and leave the internal default version in place.  But it's ONLY to be used for local development mode, _not_ in the frontend or in any production pipelines.

### Test plan for issue:

Combination of local development mode and probably deployment to INT to test `aro update-versions` in a pipeline.

### Is there any documentation that needs to be updated for this PR?

The wiki page [Multi-Version OpenShift Installations](https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki?anchor=openshift-versions&pagePath=/Build%252Dout/Multi%252DVersion%20OpenShift%20Installations) should direct readers to [RP-Config](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config) rather than [pkg/util/version/const.go](https://github.com/Azure/ARO-RP/blob/master/pkg/util/version/const.go) for updating install streams.

### CODE REVIEW

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go
Comment (jhoreman) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:53]: Hey Matt, why are we looking at 4.10 specifically here?

Comment (mbarnes) [reply pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:53]: No real reason, I just needed _some_ version to plug into `ClusterProfile` structs.

File: cmd/aro/update_ocp_versions.go
Comment (SudoBrendan) [inline cmd/aro/update_ocp_versions.go:53]: nit: would it be better to move this to a package-level const so it appears at the top of the file? Same for `INSTALLER_IMAGE_DIGESTS` below?

File: pkg/frontend/openshiftversions_list.go
Comment (SudoBrendan) [inline pkg/frontend/openshiftversions_list.go:45]: This is an interesting symptom of this change - I feel like we should have this case outlined somewhere, right? Maybe we decide that it lands in the RP release pipeline or something? We need to guarantee that we'll always return at least one version (defaultversion) to the customer at all times and prevent rollouts that would make this possible (we could also do this at an RP level, for example, we exit `1` if no default version is found, and that happens way before a request would reach `frontend`).

File: pkg/monitor/cluster/clusterversions.go
Comment (SudoBrendan) [inline pkg/monitor/cluster/clusterversions.go:59]: is there a reason we've dropped the `latestGaMinorVersion` metric?

**Review by SudoBrendan — COMMENTED** [review]
Overall this is looking pretty solid, a few questions. Thanks for the PR!

File: cmd/aro/update_ocp_versions.go
Comment (mbarnes) [reply cmd/aro/update_ocp_versions.go:53]: +0, I can do that.

File: pkg/monitor/cluster/clusterversions.go
Comment (mbarnes) [reply pkg/monitor/cluster/clusterversions.go:59]: I couldn't find any evidence of its use, and `version.DefaultInstallStream` is strictly for local development only now.

File: pkg/frontend/openshiftversions_list.go
Comment (mbarnes) [reply pkg/frontend/openshiftversions_list.go:45]: That brings up an important point: we need to run the **Update ARO OpenShift Versions** pipelines to deploy the default flag _before_ we release an RP with this change so that the scenario you're talking about can't happen.  If we're releasing the RP next week, then we should either start a version update rollout this week or hold this PR until September.

I did add a [check to "aro update-versions"](https://github.com/Azure/ARO-RP/blob/14f7393dd0c1a12bd85d79d8293bc6706701d30b/cmd/aro/update_ocp_versions.go#L60-L64) to ensure the `defaultStreams` map has one and only one item.  We can't really block RP rollouts over this but we can block OCP version updates.

Comment (mbarnes) [reply pkg/frontend/openshiftversions_list.go:45]: Additionally, the absence of a default version in CosmosDB is something I think SREs should be alerted to so the database can be corrected.  Maybe a followup task would be to add an `OpenShiftVersions` changefeed listener in the metrics code and emit a default version metric that we can craft an alert around?

File: cmd/aro/update_ocp_versions.go
Comment (mbarnes) [reply cmd/aro/update_ocp_versions.go]: It turns out we have a whole file (`/cmd/aro/const.go`) for environment variable names, but it needed a touch of rework.  See additional commits.

**Comment by mbarnes** [comment]
E2E will fail until [ARO.Pipelines: Read OpenShiftVersions from YAML configuration](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/ARO.Pipelines/pullrequest/8603681) is merged.

**Comment by mbarnes** [comment]
Needed a final commit to use the internal default OCP version for running E2E tests for CI.

**Review by SudoBrendan — DISMISSED** [review]

**Comment by mbarnes** [comment]
Placing a **Hold** on this while I test a concern I had today.  The code here is technically correct, but getting this all deployed without breaking anything may require more steps than I expected.

**Comment by mbarnes** [comment]
Confirmed my concern, unfortunately.  If we deploy the default flag to CosmosDB before the RP is aware of the flag, it will break the changefeed process in the RP frontend with the error below and the RP will fail to read valid OCP versions from CosmosDB.

```
pkg/frontend/changefeed.go:39 frontend.(*frontend).updateFromIterator() json decode error [pos 626]:
no matching struct field found when decoding stream map with key default  component=frontend
```

That means we need to make the RP aware of the default flag _before_ deploying the default flag.  So not one but _two_ RP releases:

1. Release an RP with only the default flag added to `OpenShiftVersionProperties`.  In other words, just the [1st commit](https://github.com/Azure/ARO-RP/pull/3094/commits/be25278a370c98bab995daf05e08988dcbf83611).
2. Merge this pull request and run the "Update ARO OpenShift Versions" pipeline to deploy the default version flag out to all regions.
3. Release an RP with the rest of the commits here to utilize the default version flag.

**Comment by mbarnes** [comment]
Split the 1st commit into https://github.com/Azure/ARO-RP/pull/3202.

File: pkg/util/version/const.go
Comment (petrkotas) [inline pkg/util/version/const.go:33]: I am wondering, should we also move other image versions to a database?

File: pkg/frontend/openshiftversions_list.go
Comment (petrkotas) [reply pkg/frontend/openshiftversions_list.go:45]: This is something that would not be changing continuously. I believe only simple test added to our e2e suite verifying the presence would be sufficient. Running e2e is mandatory step after rollout. This would alert us soon enough making manual corrective action possible.

**Review by petrkotas — APPROVED** [review]
Hi @mbarnes to clarify, the process of updating versions in production stays the same. The twist is we need make sure we run it before release. Otherwise, this is really nice simplification. Thank you.

File: pkg/util/version/const.go
Comment (mbarnes) [reply pkg/util/version/const.go:33]: I think it makes sense; it's configuration data.  Downsides are having to repeat the data in each of the four RP-Config cloud files, and it potentially complicates local development mode if it depends on any of those images.

**Comment by SudoBrendan** [comment]
Hey, this was tagged as `next-release` but I think we're waiting on a release of https://github.com/Azure/ARO-RP/pull/3202 before this can merge, right?

**Comment by mbarnes** [comment]
> Hey, this was tagged as next-release but I think we're waiting on a release of https://github.com/Azure/ARO-RP/pull/3202 before this can merge, right?

You're right, I guess we haven't cut a release since that merged.  So this needs to wait til **next-next-release**.

**Comment by mbarnes** [comment]
@SudoBrendan and @petrkotas had previously approved this.

File: cmd/aro/update_ocp_versions.go
Comment (SudoBrendan) [inline cmd/aro/update_ocp_versions.go:46]: Looking another time, `data` feels a little too generic... Can we make this something like `envInstallerImageDigests` or similar, or is it too late?

File: cmd/aro/update_ocp_versions.go
Comment (SudoBrendan) [inline cmd/aro/update_ocp_versions.go:42]: scope creep, I know... Feel free to :hear_no_evil: and we can groom additional work...

...but while we're touching this, can we maybe make these log messages something that would be useful in production, even adding a bit of context like "Failed to base64 decode install image digests" and use `%w` instead of `%v`?

File: cmd/aro/update_ocp_versions.go
Comment (SudoBrendan) [inline cmd/aro/update_ocp_versions.go:47]: :hear_no_evil: same scope-creep as above, make this generic error useful by including a reference to image digests, and use `%w`

**Review by SudoBrendan — APPROVED** [review]
left some comments/nits. Thanks!

File: cmd/aro/update_ocp_versions.go
Comment (mbarnes) [reply cmd/aro/update_ocp_versions.go:46]: It's meant to be generic because it's not specific to installer image digests.

I guess it's easy to miss in the diff, but this logic is now part of a helper function named `getEnvironmentData` that's used for both installer image digests and openshift versions.

File: cmd/aro/update_ocp_versions.go
Comment (mbarnes) [reply cmd/aro/update_ocp_versions.go:47]: Same answer as above, though note it does print the environment variable name first.

So it's going to be one of:
```
INSTALLER_IMAGE_DIGESTS: Failed to parse JSON: ...

OPENSHIFT_VERSIONS: Failed to parse JSON: ...
```

I can change `%v` to `%w` although I'm not sure I understand the difference.

Comment (mbarnes) [reply cmd/aro/update_ocp_versions.go]: Updated the "Refactoring" commit to change `%v` to `%w`.

**Review by kimorris27 — COMMENTED** [review]
Looks good to me! Have the changes been tested in INT yet?

**Review by jhoreman — APPROVED** [review]
Was able to personally test this in INT. Updating the default version in the cosmosDB document for OpenShiftVersions changes the default install version as expected and desired.

**Review by SudoBrendan — APPROVED** [review]

**Review by kimorris27 — APPROVED** [review]
Reviewed this back in December but am giving my official green check mark now that it's been tested in INT 🙂

File: cmd/aro/update_ocp_versions.go
Comment (mbarnes) [reply cmd/aro/update_ocp_versions.go:29]: FWIW, [ARO.Pipelines has the full struct and parser](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/ARO.Pipelines?path=/ev2/generator/rpconfig_struct.go) you're looking for.  What's happening in this context is the parser is extracting the version bits we need from RP-Config and passing it to the RP as a JSON blob by way of an environment variable.  All we're parsing here is the environment varaible, not RP-Config directly.

**Pull request merged**
